### ISSUE TITLE
"object*" is not iterable with itertools.zip_longest

### ISSUE THREAD

**Issue by jkseppan** [microsoft/pyright#5126]
**Describe the bug**
One use of `zip_longest` fails with `"object*" is not iterable` with 1.1.309 but not with 1.1.308.

**To Reproduce**

File `tmp.py`:

```py
from itertools import zip_longest

aa = [1, 2, 3]
bb = [{"num": 0, "text": "foo"}]
for a, b in zip_longest(aa, bb):
    print(a, b)
```

If I run this with Python 3.11:

```
1 {'num': 0, 'text': 'foo'}
2 None
3 None
```

If I analyze this with latest pyright:

```
$ env PYRIGHT_PYTHON_FORCE_VERSION=latest pyright ./tmp.py
/home/jks/tmp.py
  /home/jks/tmp.py:5:5 - error: "object*" is not iterable
    "__iter__" method not defined (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Expected behavior**
What I get with 1.1.308:

```
$ pyright ./tmp.py
0 errors, 0 warnings, 0 informations
```

**Screenshots or Code**

```py
from itertools import zip_longest

aa = [1, 2, 3]
bb = [{"num": 0, "text": "foo"}]
for a, b in zip_longest(aa, bb):
    print(a, b)
```

**VS Code extension or command-line**

Command-line:

```
$ pyright --version
pyright 1.1.308
$ env PYRIGHT_PYTHON_FORCE_VERSION=latest pyright --version
pyright 1.1.309
```

**Comment by jkseppan** [comment]
Happens also with `zip`, so here's a shorter test:

```py
aa = [1, 2, 3]
bb = [{"num": 0, "text": "foo"}]
for a, b in zip(aa, bb):
    print(a, b)
```

**Comment by erictraut** [comment]
Thanks for the bug report, and apologies for the regression. This will be addressed in the next release.

The bug affected the unusual case where a generic class contained a `__new__` method that used function-scoped type variables but not class-scoped type variables _and_ the class contained no `__init__` method. This combination was not represented in pyright's test matrix which allowed the regression to go undetected. I've fixed the problem and plugged the hole in the tests.
